contracts/Staking.sol

```diff
@@ -0,0 +1,559 @@
+    require(amount >= minDelegation, "invalid delegate amount");
+    require(msg.value >= amount.add(oracleRelayerFee), "not enough msg value");
+
+    // native bnb decimals is 8 on BBC, while the native bnb decimals on BSC is 18
+    uint256 convertedAmount = amount.div(TEN_DECIMALS);
+    uint256 _oracleRelayerFee = (msg.value).sub(amount);
+
+    bytes[] memory elements = new bytes[](3);
+    elements[0] = msg.sender.encodeAddress();
+    elements[1] = validator.encodeAddress();
+    elements[2] = convertedAmount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(CROSS_STAKE_CHANNELID, _RLPEncode(EVENT_DELEGATE, msgBytes), _oracleRelayerFee.div(TEN_DECIMALS));
+    payable(TOKEN_HUB_ADDR).transfer(msg.value);
+
+    delegated[msg.sender] = delegated[msg.sender].add(amount);
+    delegatedOfValidator[msg.sender][validator] = delegatedOfValidator[msg.sender][validator].add(amount);
+
+    emit delegateSubmitted(msg.sender, validator, amount, _oracleRelayerFee);
+  }
+
+  function undelegate(address validator, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(block.timestamp >= pendingUndelegateTime[msg.sender][validator], "pending undelegation exist");
+    if (amount < minDelegation) {
+      require(amount == delegatedOfValidator[msg.sender][validator], "invalid undelegate amount");
+    }
+    delegatedOfValidator[msg.sender][validator] = delegatedOfValidator[msg.sender][validator].sub(amount, "not enough funds to undelegate");
+
+    // native bnb decimals is 8 on BBC, while the native bnb decimals on BSC is 18
+    uint256 convertedAmount = amount.div(TEN_DECIMALS);
+    uint256 _oracleRelayerFee = msg.value;
+    require(_oracleRelayerFee >= oracleRelayerFee, "not enough relayer fee");
+
+    bytes[] memory elements = new bytes[](3);
+    elements[0] = msg.sender.encodeAddress();
+    elements[1] = validator.encodeAddress();
+    elements[2] = convertedAmount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(CROSS_STAKE_CHANNELID, _RLPEncode(EVENT_UNDELEGATE, msgBytes), _oracleRelayerFee.div(TEN_DECIMALS));
+    payable(TOKEN_HUB_ADDR).transfer(_oracleRelayerFee);
+
+    delegated[msg.sender] = delegated[msg.sender].sub(amount);
+    pendingUndelegateTime[msg.sender][validator] = block.timestamp.add(8*24*3600);
+
+    emit undelegateSubmitted(msg.sender, validator, amount, _oracleRelayerFee);
+  }
+
+  function redelegate(address validatorSrc, address validatorDst, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(validatorSrc != validatorDst, "invalid redelegation");
+    require(block.timestamp >= pendingRedelegateTime[msg.sender][validatorSrc][validatorDst], "pending redelegation exist");
+    require(amount >= minDelegation, "invalid redelegate amount");
```

Comment:
Otherwise, it can be all the funds of staked BNB.

Reply:
We omitted this feature